test/src/e2e_vm_tests/test_programs/should_pass/language/intrinsics/src/main.sw

```diff
@@ -0,0 +1,14 @@
+script;
+
+use std::assert::assert;
+
+fn main() -> u64 {
+
+  assert(true == true);
+  assert(true != false);
+
+  assert(__eq(1, 22) == (1 == 22));
+  assert(__eq(1, 1) == (1 == 1));
+
+  2
+}
```

Comment:
Can you be a bit more thorough here please?  We need more testing in general I think:
- This is only testing `__eq` so the test name should probably be `should_pass/language/eq_intrinsic`.
- Confirm that we can pass all the built-in types to `__eq` without type error.
- Confirm (in a `should_fail`) that you can't pass custom types (structs, enums) or all by-ref types (tuples, strings, B256, etc).
- Now that intrinsics are generalised, does the parser always parse `__xxx` calls as intrinsic calls?  What happens if you try a `___xxx()` call for an unknown intrinsic?  I think we already check for this so there might be an existing E2E test, but please check that we get a coherent error message.
- We should have an IR test (under `sway-core/tests`), just a small one confirming `__eq` generates `cmp eq` and not a `call`.  

We should also document these somewhere as we add them.  I'm not sure if they belong in the book... maybe not as they're only meant to be used by library devs.  @sezna @mohammadfawaz Ideas?

Reply:
I've addressed all these comments, except for

> What happens if you try a ___xxx() call for an unknown intrinsic? 

There already are tests that address this, such as `double_underscore_fn`.

>We should also document these somewhere as we add them.

I'll wait for @sezna @mohammadfawaz to respond.

Reply:
Ref: https://github.com/FuelLabs/sway/issues/855#issuecomment-1170054431 as a starting point.

More generally, I've suggested in the past the idea of introducing a "Sway Reference" book that is similar to the ["Rust Reference"](https://doc.rust-lang.org/reference/) which would include the language spec (we don't have `sway.pest` anymore so we don't have that in any form) as well as all the language features that are not necessarily useful for end users but only for library authors. This is a big undertaking but probably worth investing the time into long term. It may even include some compiler internals such as how monomorphization works for example.